impl/maven-impl/src/main/java/org/apache/maven/impl/DefaultPluginXmlFactory.java

```diff
@@ -38,65 +35,61 @@
 import org.apache.maven.plugin.descriptor.io.PluginDescriptorStaxReader;
 import org.apache.maven.plugin.descriptor.io.PluginDescriptorStaxWriter;
 
-import static org.apache.maven.impl.ImplUtils.nonNull;
 import static org.apache.maven.impl.StaxLocation.getLocation;
 import static org.apache.maven.impl.StaxLocation.getMessage;
 
 @Named
 @Singleton
 public class DefaultPluginXmlFactory implements PluginXmlFactory {
+
     @Override
     public PluginDescriptor read(@Nonnull XmlReaderRequest request) throws XmlReaderException {
-        nonNull(request, "request");
-        Path path = request.getPath();
-        URL url = request.getURL();
-        Reader reader = request.getReader();
-        InputStream inputStream = request.getInputStream();
-        if (path == null && url == null && reader == null && inputStream == null) {
-            throw new IllegalArgumentException("path, url, reader or inputStream must be non null");
-        }
+        return read(request.assertReadable(), setAddDefaultEntities(request, new PluginDescriptorStaxReader()));
+    }
+
+    @Override
+    public void write(XmlWriterRequest<PluginDescriptor> request) throws XmlWriterException {
+        write(request.assertWritable(), new PluginDescriptorStaxWriter());
```

Comment:
sorry, no. We are not writing the asertWritable. That is a separate operation. assert is one thing and request is another. Only the latter should be written. This is a confusing pattern, and we shouldn't use it. 

Yes, I know other code uses this. That code is poorly written and we shouldn't copy it.

Reply:
yes